Improved entry point of the documentation. (#162)
polycube-network/polycube: Documentation/intro.rst

Introduction to Polycube
========================

Polycube brief
--------------

``Polycube`` is an **open source** software framework that provides **fast** and **lightweight** **network functions**. Already available services, called `cubes`, include `bridge`, `router`, `nat`, `load balancer`, `firewall`, `DDoS mitigator`, and more.

Cube can be composed to build arbitrary **service chains** and provide custom network connectivity to **namespaces**, **containers**, **virtual machines**, and **physical hosts**.

Cubes are extremely **efficient** because are based on the recent `BPF` and `XDP` Linux kernel technologies. In addition, cubes are easily **extensible** and **customizable**.

Polycube networks can be controlled by a coherent, uniform and simple command line, named `polycubectl`, which interacts with a REST-based daemon, `polycubed`.

> Polycube provides also two proof-of-concept **standalone applications** based upon this framework.

Reviewer: `proof-of-concept` -> `proofs-of-concept`

Author: Not sure it's correct. The plural refers to applications. Let's wait some native speaker to take care of this.

`Pcn-k8s` is a brand new network plug-in for *Kubernetes*, which can handle the network of an entire datacenter and that outperforms existing solutions in terms of throughput.

> `Pcn-iptables` is a more efficient and scalable clone of *iptables*.

Reviewer 2: ```suggestion
`PCN-iptables` is a more efficient and scalable clone of existing Linux *iptables*.
```

Author: Used 'lowercase' in the rest of the text; changed to 'pcn-iptables'.

A brief overview of the Polycube layered structure, including the command line interface (CLI), standalone applications, and some of the available cubes, is shown in the picture below.

.. image:: images/polycube-archi.png
   :alt: Polycube architecture brief

Main features
-------------

Extremely fast
~~~~~~~~~~~~~~
Polycube enables extremely fast and efficient network services, thanks to its capability to run inside the Linux kernel and, whenever possible, as close as possible to the network interface card driver, which reduces the time spent in ancillary components. 

> Born for service chaining

Reviewer 2: Change of word for `Born`

~~~~~~~~~~~~~~~~~~~~~~~~~
Polycube supports the definition of multiple network scenarios through the composition of many elementary building blocks (i.e., *cubes*),  which can be combined (e.g., attached one to the other) to create complex network services.
For instance, several dockers can communicate through a bridge, which is then attached to a router to provide internet connectivity (possibly through a nat), while a firewall protects the entire infrastructure.

Polycube has been designed to simplify service chaining: cubes can be dynamically instantiated and seamlessly connected together using virtual links, mimicking traditional networks in which dedicated middlebox are connected with each other through physical wires.
As a consequence, cubes can be composed to build arbitrary service chains and provide custom network connectivity to namespaces, docker, virtual machines, and physical hosts.

Production-grade network services
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Polycube greatly simplifies the way data, control, and management planes communicate, hence enabling the creation of rich network services that include all the above features.

Reviewer 2: ```suggestion
Polycube greatly simplifies the data, control, and management planes communication, hence it enables the creation of rich network services which includes all the above features.
```

In turn, this offers a simplified environment to service developers, who can leverage the power of Polycube to write the data/control and management parts of their network services.

In detail, most network services include a *data plane*, such as the longest prefix match algorithm in a router, a *control plane*, e.g., where routing protocols are executed, and a *management plane*, devoted to the configuration and monitoring of the service.
`bpf` aims at the creation of fast data planes, leaving the rest under the responsibility of the developer; Polycube overcomes this limitation with a rich set of primitives natively provided by the framework.

> In addition, Polycube provides the software infrastructure required to overcome possible limitations of BPF in the data plane (e.g., the limited size of data plane programs), enabling the steering of packets that require complex processing in user-space, where previous limitations doesn't apply.

Reviewer: limitations `doesn't` -> `don't` apply

Reviewer 2: It should be fine, third-person speech used here.
